chain/src/chain.rs

```diff
@@ -281,52 +244,223 @@ impl<CI: ChainIndex + 'static> ChainService<CI> {
             self.shared.store().insert_transaction_address(
                 batch,
                 &hash,
-                &block.commit_transactions(),
+                block.commit_transactions(),
             );
         }
 
-        let mut hash = block.header().parent_hash().clone();
-        number -= 1;
+        for n in new_number..old_number {
+            self.shared.store().delete_block_hash(batch, n + 1);
+        }
+    }
+
+    fn get_forks(
+        &self,
+        tip_number: BlockNumber,
+        block: &Block,
+        exts: &mut Vec<(H256, BlockExt)>,
+        ext: BlockExt,
+    ) -> (bool, Vec<Block>, Vec<Block>) {
+        let mut number = block.header().number() - 1;
+        let mut old_blocks = Vec::new();
+        let mut new_blocks = Vec::new();
+
+        // The old fork may longer than new fork
+        if number < tip_number {
+            for n in number..=tip_number {
+                let hash = self.shared.block_hash(n).unwrap();
+                let old_block = self.shared.block(&hash).unwrap();
+
+                old_blocks.push(old_block);
+            }
+        }
+
+        let mut verified = ext.valid.is_some();
+        let mut hash = block.header().hash();
+
+        if !verified {
+            exts.push((hash, ext));
+        }
+
+        //TODO: remove this clone
+        new_blocks.push(block.clone());
+
+        hash = block.header().parent_hash().clone();
+
         loop {
             if let Some(old_hash) = self.shared.block_hash(number) {
                 if old_hash == hash {
+                    verified = true;
```

Comment:
I still not quite understand what you mean `verified` here. If the new fork contains only pending verification blocks, the loop finally stops at the common ancestor and will return `verified` as true.

Reply:
skip the meaning of verified, I will change the name. It just meaning no error happened.